deepdrivemd/agents/stream/dbscan.py

```diff
@@ -0,0 +1,698 @@
+        latent_dim=cfg.latent_dim,
+    )
+    cvae.load(model_path)
+    return cvae
+
+
+def wait_for_model(cfg: OutlierDetectionConfig) -> str:
+    """Wait for the trained model to be published by machine learning pipeline.
+
+    Returns
+    -------
+    str
+         path to the model
+    """
+
+    while True:
+        if os.path.exists(cfg.best_model):
+            break
+        print(f"No model {cfg.best_model}, sleeping")
+        sys.stdout.flush()
+        time.sleep(cfg.timeout2)
+    return cfg.best_model
+
+
+def wait_for_input(cfg: OutlierDetectionConfig) -> List[str]:
+    """Wait for enough data to be produced by simulations.
+
+    Returns
+    -------
+    List[str]
+         a list of aggregated bp file
+    """
+    while True:
+        bpfiles = glob.glob(str(cfg.agg_dir / "*/*/agg.bp"))
+        if len(bpfiles) == cfg.num_agg:
+            break
+        print(f"Waiting for {cfg.num_agg} agg.bp files")
+        time.sleep(cfg.timeout1)
+
+    print(f"bpfiles = {bpfiles}")
+
+    # Wait for enough time steps in each bp file
+    while True:
+        enough = True
+        for bp in bpfiles:
+            com = f"bpls {bp}"
+            a = subprocess.getstatusoutput(com)
+            if a[0] != 0:
+                enough = False
+                print(f"Waiting, a = {a}, {bp}")
+                break
+            try:
+                steps = int(a[1].split("\n")[0].split("*")[0].split(" ")[-1])
+            except Exception as e:
+                print("Exception ", e)
+                steps = 0
+                enough = False
+            if steps < cfg.min_step_increment:
+                enough = False
+                print(f"Waiting, steps = {steps}, {bp}")
+                break
+        if enough:
+            break
+        else:
+            time.sleep(cfg.timeout2)
+
+    return bpfiles
+
+
+def dirs(cfg: OutlierDetectionConfig) -> Tuple[str, str, str]:
+    """Create tmp_dir and published_dir into which outliers are written
+
+    Returns
+    -------
+    Tuple[str, str, str]
+         paths to top, tmp and published directories
+    """
+    top_dir = cfg.output_path
+    tmp_dir = f"{top_dir}/tmp"
+    published_dir = f"{top_dir}/published_outliers"
+
+    if not os.path.exists(tmp_dir):
+        os.mkdir(tmp_dir)
+    if not os.path.exists(published_dir):
+        os.mkdir(published_dir)
+    return top_dir, tmp_dir, published_dir
+
+
+def predict(
+    cfg: OutlierDetectionConfig,
+    model_path: str,
+    cvae_input: Tuple[
+        List[np.ndarray],
+        List[np.ndarray],
+        List[np.ndarray],
+        List[np.ndarray],
+        List[np.ndarray],
+    ],
+    batch_size: int = 32,
+) -> np.ndarray:
+    """Project contact maps into the middle layer of CVAE
+
+    Parameters
+    ----------
+    cfg : OutlierDetectionConfig
+    model_path : str
+    cvae_input : Tuple[List[np.ndarray], List[np.ndarray], List[np.ndarray], List[np.ndarray], List[np.ndarray]]
+    batch_size : int
+
```

Comment:
Docstring descriptions for input variables and return.

Reply:
Done